### PR TITLE
feature/v-treeview

### LINKED ISSUE #2600: [Feature Request] The tree view component is so necessary!
### New Functionality
Many other vue component libraries has this component!
like
-  `element ui` http://element-cn.eleme.io/#/en-US/component/tree
- `iview` https://www.iviewui.com/components/tree-en
and another material vue framework `quasar framework`
http://quasar-framework.org/components/tree.html
can do this work for vuetify? thx

### Improvements
I use v-list do this current,but is not perfect plan

### Bugs or Edge Cases it Helps Avoid
will generator tree-view component fast and perfect

### PR DESCRIPTION
**Pull request by nekosaur**
## Description
v-treeview component

## Todo
- [x] more tests

## Motivation and Context
closes #2600 

## How Has This Been Tested?
playground, tests

## Markup:

<details>

```vue
<template>
  <v-app>
    <v-content>
      <v-container fluid grid-list-xl>
        <v-layout row wrap>
          <v-flex xs6>
            Default
            <v-treeview :items="items"></v-treeview>
          </v-flex>
          <v-flex xs6>
            Selectable
            <v-treeview :items="items" selectable :selected.sync="selected"></v-treeview>
          </v-flex>
          <v-flex xs6>
            Transition
            <v-treeview :items="items" transition></v-treeview>
          </v-flex>
          <v-flex xs6>
            Slots prepend & append
            <v-treeview :items="items">
              <v-icon slot="prepend" slot-scope="props">{{ props.item.icon }}</v-icon>
              <div class="d-flex grow justify-end" slot="append" slot-scope="props">
                <v-btn class="shrink" v-if="props.active" small icon @click.stop="remove(items, props.item.id)">
                  <v-icon color="error">delete</v-icon>
                </v-btn>
              </div>
            </v-treeview>
          </v-flex>
          <v-flex xs6>
            Multiple active
            <v-treeview :items="items" multiple-active></v-treeview>
          </v-flex>
          <v-flex xs6>
            Dynamic loading of children
            <v-treeview :items="dynamic" :load-children="loadChildren"></v-treeview>
          </v-flex>
          <v-flex xs12>
            <v-btn @click="addRootTree">add another root</v-btn>
          </v-flex>
        </v-layout>
      </v-container>
    </v-content>
  </v-app>
</template>

<script>

let id = 0

function generateTree (depth = 2, minChildren = 2, maxChildren = 4, parent = null) {
  const node = {
    id: ++id,
    icon: 'home',
    name: 'whatever'
  }

  if (depth === 0) return node

  const childCount = Math.floor(Math.random() * (maxChildren - minChildren + 1) + minChildren)

  if (childCount > 0) node.children = []

  for (let i = 0; i < childCount; i++) {
    node.children.push(generateTree(depth - 1, minChildren, maxChildren, node))
  }

  return node
}

const pause = (ms) => new Promise(resolve => setTimeout(resolve, ms))

export default {
  methods: {
    addRootTree () {
      this.items.push(generateTree())
    },
    findNode (items, id) {
      if (id === null || !items) return null

      for (let i = 0; i < items.length; i) {
        const item = items[i]

        if (item.id === id) return item

        const found = this.findNode(item.children, id)

        if (found) return found
      }

      return null
    },
    async loadChildren (item) {
      await pause(1000)
      id += 1
      item.children = [{
        id,
        name: 'Generated child',
        children: []
      }]
    },
    remove (items, id) {
      for (let i = 0; i < items.length; i++) {
        const item = items[i]

        if (item.id === id) {
          items.splice(i, 1)
          return true
        }

        if (item.children) {
          if (this.remove(item.children, id) && !item.children.length) {
            delete item.children
          }
        }
      }
    }
  },
  data: () => ({
    selected: [],
    items: [generateTree()],
    dynamic: [
      { id: 0, name: 'Root', children: [] }
    ]
  })
}
</script>
```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any feature but make things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes, `dev` for new features and breaking changes).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have created a PR in the [documentation](https://github.com/vuetifyjs/vuetifyjs.com) with the necessary changes.

### CODE REVIEW

File: src/components/VTreeview/VTreeviewNode.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        props: {
+          color: this.isSelected ? this.selectedColor : undefined
+        },
+        on: {
+          click: async (e: MouseEvent) => {
+            e.stopPropagation()
+
+            if (this.isLoading) return
+
+            await this.checkChildren()
```
[the remark this answers is not in the document]
Comment (nekosaur) [reply src/components/VTreeview/VTreeviewNode.ts]: It should get transformed by babel. I think.

[the remark this answers is not in the document]
Comment (KaelWD) [reply src/components/VTreeview/VTreeviewNode.ts]: We aren't using it because it transpiles to a massive pile of iterators.

[the remark this answers is not in the document]
Comment (nekosaur) [reply src/components/VTreeview/VTreeviewNode.ts]: Alright

**Comment by johnleider** [comment]
Pre-defining values in the **active** prop has no effect

**Comment by LeoMartinDev** [comment]
Are drag&drop capabilities planed ? It would be a great feature IMO.

**Comment by nekosaur** [comment]
Please do not comment on closed issues or PRs. If you have a question, please come to our [chat](https://chat.vuetifyjs.com).

There are currently no plans to add drag & drop to treeview.
